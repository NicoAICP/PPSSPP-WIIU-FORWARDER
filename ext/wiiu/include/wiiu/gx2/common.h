#pragma once

#define GX2_SCAN_BUFFER_ALIGNMENT      0x1000
#define GX2_SHADER_ALIGNMENT           0x100
#define GX2_CONTEXT_STATE_ALIGNMENT    0x100
#define GX2_UNIFORM_BLOCK_ALIGNMENT    0x100
#define GX2_DISPLAY_LIST_ALIGNMENT     0x20
#define GX2_VERTEX_BUFFER_ALIGNMENT    0x40
#define GX2_INDEX_BUFFER_ALIGNMENT     0x20

#define GX2_ENABLE                     TRUE
#define GX2_DISABLE                    FALSE

#define GX2_TRUE                       TRUE
#define GX2_FALSE                      FALSE

#ifndef GX2_COMP_SEL
#define GX2_COMP_SEL(c0, c1, c2, c3) (((c0) << 24) | ((c1) << 16) | ((c2) << 8) | (c3))
#ifdef __cplusplus
#include <functional>
#endif
#include <stdio.h>

#define _x 0
#define _y 1
#define _z 2
#define _w 3
#define _r 0
#define _g 1
#define _b 2
#define _a 3
#define _0 4
#define _1 5
#endif


#define GX2_DISABLE_WRAPS
