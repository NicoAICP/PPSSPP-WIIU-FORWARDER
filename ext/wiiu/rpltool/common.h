#ifndef COMMON_H
#define COMMON_H

#include <ctype.h>
#include <stdint.h>
#include <stdio.h>

typedef int8_t s8;
typedef int16_t s16;
typedef int32_t s32;

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;

#define GET_ENUM_NAME(prefix, name, val) \
   case prefix##name: \
      return #name;

#define EMIT_ENUM_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(name type) { \
      switch (type) { cases } \
      return "unknown"; \
   }

#define GET_FLAG_NAME(prefix, name, val) \
   case prefix##name: \
      pos += snprintf(buffer + pos, sizeof(buffer) - pos, "%s%s", #name, seperator); \
      flags &= ~(1u << i); \
      break;

#define EMIT_FLAGS_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(int flags, const char *seperator) { \
      int i; \
      static char buffer[32 * 32]; \
      int pos = 0; \
      for (i = 0; i < 32; i++) \
         if ((1u << i) & flags) \
            switch (1u << i) { cases } \
      if (flags) \
         snprintf(buffer + pos, sizeof(buffer) - pos, "0x%08X", flags); \
      else if (pos) \
         buffer[pos - strlen(seperator)] = '\0'; \
      else \
         return "NULL"; \
      return buffer; \
   }

#define MAKE_MAGIC(a, b, c, d) (((a) << 24) | ((b) << 16) | ((c) << 8) | ((d) << 0))

#define log_var(X) printf("%-35s: 0x%0*X (%u)\n", #X, sizeof(X) * 2, (u32)(X), (u32)(X))
#define log_svar(X) printf("%-35s: 0x%0*X (%i)\n", #X, sizeof(X) * 2, (u32)(X), (s32)(X))
#define log_array(X) debug_array(X, sizeof(X), #X)
#define log_enum(type, X) printf("%-35s: 0x%X (%s)\n", #X, (X), type##_to_str(X))
#define log_flags(type, X) printf("%-35s: 0x%08X (%s)\n", #X, (X), type##_to_str(X, " | "))
#define log_magic(X) \
   printf("%-35s: '%c''%c''%c''%c' (0x%08X)\n", #X, (u32)(X) >> 24, (u32)(X) >> 16, (u32)(X) >> 8, (u32)(X), (u32)(X))
#define log_carray(X) debug_char_array(X, sizeof(X), #X)
#define log_str(X) printf("%-35s: \"%s\"\n", #X, (char *)(X))
#define log_off_str(X, data) printf("%-35s: 0x%08X \"%s\"\n", #X, (X), (char *)data + (X))
#define log_off_strs(X, data) \
   do { \
      printf("%-35s: 0x%08X ", #X, (X)); \
      const char *ptr = (char *)data + (X); \
      while (*ptr) { \
         ptr += printf("\"%s\" ", ptr) - 2; \
      } \
      printf("\n"); \
   } while (0)

static inline void dump_array(uint32_t *src, int size);
static inline void debug_array(uint32_t *src, int size, const char *name) {
   printf("%-35s:", name);
   dump_array(src, size);
}
static inline void debug_char_array(uint8_t *src, int size, const char *name) {
   int i;
   printf("%-35s: ", name);
   //   for (i = 0; i < size; i++)
   //      printf("%-2c ", isalnum(src[i])?src[i]: '?');
   //   printf("\n%-35s: ", "");
   for (i = 0; i < size; i++)
      printf("%02X ", src[i]);
   printf("\n");
}

static inline void dump_array(uint32_t *src, int size) {
   int i;
   size = (size / 4) - 1;
   printf("\n{");
   for (i = 0; i < size; i++) {
      if (!(i & 3))
         printf("\n   ");

      printf("0x%08X, ", src[i]);
   }

   if (!(i & 3))
      printf("\n   ");

   printf("0x%08X\n}\n", src[i]);
}

static inline u32 align_up(u32 x, u32 align) { return (x + (align - 1)) & ~(align - 1); }
static inline u32 align_down(u32 x, u32 align) { return x & ~(align - 1); }

#endif // COMMON_H
