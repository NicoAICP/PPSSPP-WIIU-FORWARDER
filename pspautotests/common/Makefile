TARGET_LIB = libcommon.a
EXTRA_TARGETS = libcommon_kernel.a
OBJS = common.o vram.o sysmem-imports.o
KERNEL_OBJS = common_kernel.o vram.o sysmem-imports.o

INCDIR = 
CFLAGS = -O0 -G 0 -g -Wall
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
ASFLAGS = $(CFLAGS)

LIBDIR =
LDFLAGS = -G0

PSPSDK=$(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak

%.o: %.S
	$(AS) $(ASFLAGS) -c -o $@ $<

common_kernel.o: common.c
	$(CC) $(CFLAGS) -DCOMMON_KERNEL=1 -c -o $@ $<

libcommon_kernel.a: $(KERNEL_OBJS)
	$(AR) cru $@ $(KERNEL_OBJS)
	$(RANLIB) $@
