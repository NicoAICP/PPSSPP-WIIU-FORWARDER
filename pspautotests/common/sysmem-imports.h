extern int sceKernelSetCompiledSdkVersion(int ver);
extern int sceKernelSetCompiledSdkVersion370(int ver);
extern int sceKernelSetCompiledSdkVersion380_390(int ver);
extern int sceKernelSetCompiledSdkVersion395(int ver);
extern int sceKernelSetCompiledSdkVersion401_402(int ver);
extern int sceKernelSetCompiledSdkVersion500_505(int ver);
extern int sceKernelSetCompiledSdkVersion507(int ver);
extern int sceKernelSetCompiledSdkVersion600_602(int ver);
extern int sceKernelSetCompiledSdkVersion603_605(int ver);
extern int sceKernelSetCompiledSdkVersion606(int ver);
extern int sceKernelGetCompiledSdkVersion();