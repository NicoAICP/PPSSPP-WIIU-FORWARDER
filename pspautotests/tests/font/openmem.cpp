#include <common.h>
#include <pspmodulemgr.h>
#include <pspiofilemgr.h>
#include <malloc.h>
extern "C" {
#include "libfont.h"
}

static FontLibraryHandle fontLib;
static FontLibraryHandle fontLibDone;
static int g_allocated = 0;
static int g_fontFile = 0;
static void *g_fontData = NULL;
static u32 g_fontDataSize = 0;

static void *allocQuiet(void *arg, u32 size) {
	g_allocated += size;
	void *p = malloc(size);
	//checkpoint("%08x = alloc(0x%x)", p, size);
	return p;
}

static void freeQuiet(void *arg, void *p) {
	//checkpoint("free(%08x)", p);
	free(p);
}

// Returns opaque handle, always set errorCode.
static u32 openFunc(void *arg, const char *filename, int *errorCode) {
	checkpoint("   open(%08x, %s, &errorCode)", (uint)arg, filename);
	g_fontFile = sceIoOpen(filename, PSP_O_RDONLY, 0);
	*errorCode = g_fontFile < 0 ? -1 : 0;
	return 0x1337;
}

// Return error code on failure.
static int closeFunc(void *arg, u32 handle) {
	checkpoint("   close(%08x, %08x)", (uint)arg, (uint)handle);
	return sceIoClose(g_fontFile);
}

// Return sizes read, always set errorCode.
static u32 readFunc(void *arg, int handle, void *ptr, u32 size, u32 count, int *errorCode) {
	checkpoint("   read(%08x, %08x, ptr, %08x, %08x, &errorCode)", (uint)arg, (uint)handle, (uint)size, (uint)count);
	u32 res = sceIoRead(g_fontFile, ptr, size * count);
	if (res < size * count) {
		*errorCode = -1;
		return 0;
	}
	*errorCode = 0;
	return res / size;
}

// Return 0 on success, other on error.
static int seekFunc(void *arg, int handle, u32 pos) {
	checkpoint("   seek(%08x, %08x, %08x)", (uint)arg, (uint)handle, (uint)pos);
	int res = sceIoLseek32(g_fontFile, pos, PSP_SEEK_SET);
	if (res < 0) {
		return -1;
	}
	return 0;
}

static int errorFunc(void *arg, u32 b, u32 c, u32 d, u32 e, u32 f, u32 g, u32 h) {
	checkpoint("   error(%08x, %08x, %08x, %08x, %08x, %08x, %08x, %08x)", (uint)arg, (uint)b, (uint)c, (uint)d, (uint)e, (uint)f, (uint)g, (uint)h);
	return -1;
}

static int ioFinishFunc(void *arg, u32 b, u32 c, u32 d, u32 e, u32 f, u32 g, u32 h) {
	checkpoint("   ioFinish(%08x, %08x, %08x, %08x, %08x, %08x, %08x, %08x)", (uint)arg, (uint)b, (uint)c, (uint)d, (uint)e, (uint)f, (uint)g, (uint)h);
	return -1;
}

bool loadFontModule() {
	checkpointNext("Init");
	if (RUNNING_ON_EMULATOR) {
		return true;
	}
	SceUID fontModule = sceKernelLoadModule("libfont.prx", 0, NULL);
	if (fontModule <= 0) {
		printf("TEST ERROR: Unable to load libfont.prx\n");
		return false;
	}

	int status = -1;
	int result = sceKernelStartModule(fontModule, 0, NULL, &status, NULL);
	if (result != fontModule || status != 0) {
		printf("TEST ERROR: libfont.prx startup failed (%08x, %08x)\n", result, status);
		return false;
	}
	return true;
}

void testOpen(const char *title, FontLibraryHandle lib, const void *data, u32 size, bool useError = true) {
	uint error = -1;
	g_allocated = 0;
	FontHandle font = sceFontOpenUserMemory(lib, data, size, useError ? &error : NULL);
	if (font <= 0) {
		checkpoint("%s: Failed (%08x, %08x)", title, font, error);
	} else if (error != 0) {
		checkpoint("%s: Error (%08x, %08x)", title, font, error);
	} else {
		int allocated = g_allocated;
		sceFontClose(font);
		checkpoint("%s: OK (allocated %d)", title, allocated);
	}
}

void testBasicParams() {
	checkpointNext("sceFontOpenUserMemory libs:");
	testOpen("  Normal", fontLib, g_fontData, g_fontDataSize, true);
	testOpen("  Invalid", 0, g_fontData, g_fontDataSize, true);
	// Crashes.
	//testOpen("  Done", fontLibDone, g_fontData, g_fontDataSize, true);

	checkpointNext("sceFontOpenUserMemory pointers:");
	testOpen("  NULL with no size", fontLib, NULL, 0, true);
	testOpen("  NULL with size", fontLib, NULL, g_fontDataSize, true);

	checkpointNext("sceFontOpenUserMemory sizes:");
	// 1 will crash, for example...
	const int sizes[] = {-1, 0, g_fontDataSize + 1, g_fontDataSize * 2};
	for (size_t i = 0; i < ARRAY_SIZE(sizes); ++i) {
		char temp[64];
		snprintf(temp, sizeof(temp), "  %d", sizes[i]);
		testOpen(temp, fontLib, g_fontData, sizes[i], true);
	}

	// Crashes.
	//checkpointNext("sceFontOpenUserMemory errors:");
	//testOpen("  Normal", fontLib, g_fontData, g_fontDataSize, false);
}

void testTooMany() {
	FontHandle fonts[4];

	checkpointNext("sceFontOpenUserMemory too many:");

	uint error;
	for (int i = 0; i < 4; ++i) {
		fonts[i] = sceFontOpenUserMemory(fontLib, g_fontData, g_fontDataSize, &error);
		if (fonts[i] == 0 || error != 0) {
			checkpoint("  Filling fonts failed: %08x, %08x", fonts[i], error);
		}
	}

	int result = sceFontOpenUserMemory(fontLib, g_fontData, g_fontDataSize, &error);
	checkpoint("  5th font: %08x, %08x", result, error);
	
	for (int i = 0; i < 4; ++i) {
		sceFontClose(fonts[i]);
	}
}

void testCallbacks() {
	fontLib->params.openFuncAddr = openFunc;
	fontLib->params.closeFuncAddr = closeFunc;
	fontLib->params.readFuncAddr = readFunc;
	fontLib->params.seekFuncAddr = seekFunc;
	fontLib->params.errorFuncAddr = (u32 *)errorFunc;
	fontLib->params.ioFinishFuncAddr = (u32 *)ioFinishFunc;

	checkpointNext("sceFontOpenUserMemory callbacks:");
	testOpen("  With callbacks set", fontLib, g_fontData, g_fontDataSize, true);
}

void testRefCount() {
	uint error;
	FontCharInfo charInfo;
	FontHandle font1 = sceFontOpenUserMemory(fontLib, g_fontData, g_fontDataSize, &error);
	FontHandle font2 = sceFontOpenUserMemory(fontLib, g_fontData, g_fontDataSize, &error);
	sceFontClose(font2);
	
	checkpointNext("sceFontOpen reference counting:");
	testOpen(  "While open", fontLib, g_fontData, g_fontDataSize);
	checkpoint("  GetCharInfo on open: %08x", sceFontGetCharInfo(font2, '.', &charInfo));
	checkpoint("  GetCharInfo on closed: %08x", sceFontGetCharInfo(font2, '.', &charInfo));
	sceFontClose(font1);
	// Crashes.
	//checkpoint("  GetCharInfo on both closed: %08x", sceFontGetCharInfo(font2, '.', &charInfo));

	sceFontClose(font1);
	sceFontClose(font2);
}

extern "C" int main(int argc, char *argv[]) {
	if (!loadFontModule()) {
		return 1;
	}

	int fd = sceIoOpen("ltn0.pgf", PSP_O_RDONLY, 0);
	g_fontDataSize = sceIoLseek32(fd, 0, PSP_SEEK_END);
	sceIoLseek32(fd, 0, PSP_SEEK_SET);
	g_fontData = new u8[g_fontDataSize];
	sceIoRead(fd, g_fontData, g_fontDataSize);
	sceIoClose(fd);

	FontNewLibParams libParams;
	memset(&libParams, 0, sizeof(libParams));
	libParams.allocFuncAddr = allocQuiet;
	libParams.freeFuncAddr = freeQuiet;
	libParams.numFonts = 4;

	uint error = -1;
	fontLib = sceFontNewLib(&libParams, &error);
	fontLibDone = sceFontNewLib(&libParams, &error);
	sceFontDoneLib(fontLibDone);

	testBasicParams();
	// That messed with our lib, let's reset.
	sceFontDoneLib(fontLib);
	fontLib = sceFontNewLib(&libParams, &error);

	testTooMany();
	testCallbacks();
	testRefCount();

	sceFontDoneLib(fontLib);

	delete [] g_fontData;

	return 0;
}
