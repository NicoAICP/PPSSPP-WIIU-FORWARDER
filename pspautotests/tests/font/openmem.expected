[x] Init

[x] sceFontOpenUserMemory libs:
[x]   Normal: OK (allocated 12)
[x]   Invalid: Failed (00000000, 80460002)

[x] sceFontOpenUserMemory pointers:
[x]   NULL with no size: Failed (00000000, 80460003)
[x]   NULL with size: Failed (00000000, 80460003)

[x] sceFontOpenUserMemory sizes:
[x]   -1: OK (allocated 12)
[x]   0: Failed (00000000, 80460003)
[x]   67661: OK (allocated 12)
[x]   135320: OK (allocated 12)

[x] sceFontOpenUserMemory too many:
[x]   5th font: 00000000, 80460009

[x] sceFontOpenUserMemory callbacks:
[x]   With callbacks set: OK (allocated 12)

[x] sceFontOpen reference counting:
[x] While open: OK (allocated 0)
[x]   GetCharInfo on open: 00000000
[x]   GetCharInfo on closed: 00000000
